TGM/src/main/java/network/warzone/tgm/modules/region/SphereRegion.java

```diff
@@ -38,11 +42,23 @@ public boolean contains(Block block) {
         return contains(block.getLocation());
     }
 
+
     @Override
     public Location getCenter() {
         return this.center;
     }
 
+    @Override
+    public Location getRandomLocation() {
+        Location location = bound.getRandomLocation();
+
+        while (!contains(location)) {
```

Comment:
Instead of generating a location in a bounding cuboid and repeatedly checking for whether the point is included, it would be best to generate a pair of [spherical coordinates](https://en.wikipedia.org/wiki/Spherical_coordinate_system), and then converting these into cartesian coordinates, offsetting them by the center.

Similarly, it would be best to do the same for hemispheres (restricting the value of theta or phi depending on which axis the hemisphere is formed).

For cylinders, [cylindrical coordinates](https://mathworld.wolfram.com/CylindricalCoordinates.html) can be used, which is just polar coordinates with a height parameter.

Reply:
```
Cylinder Region (cuboid bounds): 7.664s, max = 12
Cylinder Region (trigonometry): 20.266s
Cylinder cuboids bounds are 2.64x faster
Sphere Region (cuboid bounds): 11.392s, max = 27
Sphere Region (trigonometry): 49.58s
Sphere cuboid bounds are 4.35x faster
```

These are my benchmark results with 100 million iterations. I think it's better to keep the cuboid bounds method of generating random (hemi)sphere and cylinder locations. If we only use it to generate random spawn points, yeah, it wouldn't matter at all, but if we'd ever use it for other stuff which requires a lot of random locations, it will surely be useful to save some performance. I'm thinking of minigames which spawn anvils or TNT above you where the objective is to survive or anything similar.

The max number means how often a new random cuboid location had to be generated in the worst case. Considering the worst case is only 27 in the 100 million iterations I ran, it isn't bad at all.

https://discord.com/channels/333669866735271938/346318260410777612/836505105326538772